foreshadow/core/base.py

```diff
@@ -0,0 +1,185 @@
+"""General base classes used across Foreshadow."""
+from abc import abstractmethod
+from copy import deepcopy
+
+from sklearn.pipeline import Pipeline
+
+from foreshadow.transformers.base import ParallelProcessor, SmartTransformer
+from foreshadow.utils import check_df
+
+
+class PreparerStep(Pipeline):
+    """Abstract Base class for any pipeline step of DataPreparer.
+
+    This class automatically wraps the defined pipeline to make it
+    parallelizable across a DataFrame. To make this possible and still be
+    customizable, subclasses must implement get_transformer_list and
+    get_transformer_weights, which are two inputs the ParallelProcessor
+    (the class used to make parallelization possible).
+
+
+    The transformer_list represents the mapping from columns to
+    transformers, in the form of ['name', 'transformer', ['cols']],
+    where the [cols] are the cols for transformer 'transformer. These
+    transformers should be SmartTransformers for any subclass.
+
+    The transformer_weights are multiplicative weights for features per
+    transformer. Keys are transformer names, values the weights.
+
+    """
+
+    def __init__(self, steps, *args, **kwargs):
+        """Set the original pipeline steps internally.
+
+        Takes a list of desired SmartTransformer steps and stores them as
+        self._steps. Constructs self an sklearn pipeline object.
+
+        Args:
+            steps: list of ('name', 'SmartTransformer') tuples, where the
+                latter is a class referenece.
+            *args: args to Pipeline constructor.
+            **kwargs: kwargs to PIpeline constructor.
+        """
+        self._steps = steps  # class pointers
+        super().__init__(steps, *args, **kwargs)
+
+    @staticmethod
+    def one_smart_all_cols(smart, X):
+        """Apply separate smart transformer to each column.
+
+        Args:
+            smart: the SmartTransformer to apply.
+            X: the input dataset points
+
+        Returns:
+            (transformer_list, transformer_weights) as inputs for
+            ParallelProcessor
+
+        """
+        transformer_list = [(smart.__name__, smart(), [col]) for col in X]
+        transformer_weights = None
+        return transformer_list, transformer_weights
+
+    def parallelize_smart_steps(self, X):
+        """Make self.steps for internal pipeline methods parallelized.
+
+        Takes self._steps passed at construction time and wraps each step
+        with ParallelProcessor to parallelize it across a DataFrame's columns.
+        Made possible sing get_transformer_list and get_transformer_weights
+        which must be implemented by the subclass.
+
+        Args:
+            X: DataFrame
+
+        """
+        steps = []
+        for step in self._steps:
+            name = step[0]
```

Comment:
might make sense to cast these to named tuples so that we can access them appropriately

Reply:
ie name, step